arangod/Cluster/ClusterCollectionMethods.cpp

```diff
@@ -192,9 +192,56 @@ Result impl(ClusterInfo& ci, ArangodServer& server,
   double pollInterval = ci.getPollInterval();
   AgencyCallbackRegistry& callbackRegistry = ci.agencyCallbackRegistry();
 
-  // TODO Timeout?
   std::vector<std::string> collectionNames = writer.collectionNames();
-  while (true) {
+  // Retry for at most 60s:
+  auto startTime = std::chrono::steady_clock::now();
+  while (std::chrono::steady_clock::now() - startTime <
+         std::chrono::seconds(60)) {
+    // Now check if any of the to-be-created collections has
+    // `distributeShardsLike` set to a collection, which already exists.
+    // If so, we need to check if any of its shards currently have
+    // their leader asked to resign!
+    std::unordered_set<std::string> distributeShardsLikeColls;
+    auto const& colls = writer.collectionsToCreate();
+    for (auto const& c : colls) {
+      if (c.properties().distributeShardsLike.has_value()) {
+        distributeShardsLikeColls.insert(
+            c.properties().distributeShardsLike.value());
+      }
+    }
+    // If we have some, let's see if they are about to be generated or not:
+    if (!distributeShardsLikeColls.empty()) {
+      for (auto const& c : colls) {
+        auto const& name = c.getName();
+        distributeShardsLikeColls.erase(name);
+      }
+    }
+    // If we still have some, they must be existing collections and we must
+    // check that their leaders are not asked to resign:
+    if (!distributeShardsLikeColls.empty()) {
+      std::vector<std::string> shards;
+      bool seenResignation = false;
+      for (auto const& c : distributeShardsLikeColls) {
+        auto coll = ci.getCollection(databaseName, c);
+        for (auto const& s : *coll->shardIds()) {
+          if (!s.second.empty() && s.second[0].starts_with("_")) {
+            seenResignation = true;
+          }
+          shards.push_back(std::string(s.first));
+        }
+      }
+      if (seenResignation) {
+        // In this case we must not proceed, or else the MoveShard job
+        // which is currently asking all leaders to resign is getting stuck.
+        // The outer retry loop will retry.
+        LOG_TOPIC("abbbe", INFO, Logger::CLUSTER)
+            << "Cannot create collection since a leader is being moved, "
+               "delaying for 1 second...";
+        std::this_thread::sleep_for(std::chrono::seconds(1));
+        continue;
+      }
```

Comment:
Is it assured, by an existing precondition (plan version?), that the things we checked here (no resigned dsl-leaders) will be unchanged when the transaction is being written?

Reply:
Yes, it is. We note the `/arango/Plan/Version`, which `ClusterInfo` used to provide the checked information. Then we add a precondition that this version is still the same. This should do.